File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
def solution(maps):
    answer = []
    lands = []
    dirs = [[1,0],[0,1],[-1,0],[0,-1]]
    
    n,m = len(maps), len(maps[0])
    for i in range(n):
        for j in range(m):
            if maps[i][j] == 'X':
                continue
                
            que = [(i,j)]
            cnt = int(maps[i][j])
            maps[i] = maps[i][:j] + 'X' + maps[i][j+1:]
            while que:
                a, b = que.pop()
                for d in dirs:
                    na, nb = a+d[0], b+d[1]
                    if na < 0 or na >= n or nb < 0 or nb >= m or maps[na][nb] == 'X':
                        continue
                    cnt += int(maps[na][nb])
                    maps[na] = maps[na][:nb] + 'X' + maps[na][nb+1:]
                    que.append((na,nb))
            lands.append(cnt)
    lands.sort()
    return lands if len(lands) > 0 else [-1]

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.kt
import java.util.*

class Solution {
    fun solution(maps: Array<String>): IntArray {
    var answer = mutableListOf<Int>()
    val dirs = arrayOf(Pair(0,1), Pair(1,0), Pair(0,-1), Pair(-1,0))
    val nm = maps.map { it.split("").filter { it != "" }.toMutableList() }.toMutableList()
    val n = nm.size
    val m = nm[0].size

    for(i in 0 until nm.size) {
        for(j in 0 until nm[i].size) {
            if(nm[i][j] == "X" || nm[i][j] == ".") {
                continue
            }

            var sum = nm[i][j].toInt()
            val queue: Queue<Pair<Int, Int>> = LinkedList()
            queue.add(Pair(i, j))
            nm[i][j] = "."

            while(queue.isNotEmpty()){
                val cur = queue.poll()
                for(dir in dirs){
                    val nx = cur.first + dir.first
                    val ny = cur.second + dir.second
                    if(nx < 0 || ny < 0 || nx >= n || ny >= m) continue
                    if(nm[nx][ny] == "X" || nm[nx][ny] == ".") continue
                    sum += nm[nx][ny].toInt()
                    nm[nx][ny] = "."
                    queue.add(Pair(nx, ny))
                }

            }

            answer.add(sum)
        }
    }

    return if(answer.isEmpty()) intArrayOf(-1) else answer.sorted().toIntArray()
    }
}